Overview scale options

ronso0 opened the pull request:
Fixup for #14309 as the new ReplayGain scaling was considered a regression for certain use case(s).
https://github.com/mixxxdj/mixxx/pull/14309#issuecomment-2673055066

This replaces the overview 'Normalize' checkbox with two options
* Normalize to peak
* Use Waveform "Global" gain and ReplayGain (if enabled)

The second option uses the same gains/factors as the scrolling waveforms, meaning it uses settings in the Waveforms and Normalization pages.
Making 'not normalize' an explicit option hopefully clarifies the situation.

The overviews in the library are now always normalized.
_____
<sub>Initial concept so the conversation here can be understood</sub>
~~This adds 3 overview gain modes~~
* ~~File gain (what we had before this PR with 'Normalize' unchecked) + factor spinbox~~
* ~~Normalize to peak~~
* ~~waveform "All" gain + ReplayGain~~

ronso0 commented:
Can someone take a look?
As mentioned in the description, it looks like the previous change (use ReaplyGain) is considered a bug by some users.

ronso0 commented:
Well, I tried to stick with the terms we already use in the preferences, which are "ReplayGain" in Normalization preferences and "Normalize", the previous overview checkbox.

But we may switch to
* Original File Gain
* Peak Normalized
* Waveform "All" Gain incl. ReplayGain
* Fixed Gain [spinbox]

Or, appending the custom factor to file gain:
* Original File Gain, custom Factor [spinbox]
* Peak Normalized
* Waveform "All" Gain incl. ReplayGain

> Fixed gain [spinbox] (do we want to have that in DB?) for 0 at unity? Both works for me.

I'm not sure what you are suggesting, can you explain please?
IMO this is a user setting that should not interfere with the db.

daschuer commented:
Currently the spin box cintains a factor with 1 at unity. 
Normally gain values are expressed as dB with 0 at unity.

daschuer commented:
I vote for three rows, because than it is clear that the custom gain is applied to the file samples. 

"Original File Gain" there is nothing like that. We have "flies samples" in the file without a gain and apply now one. So it may be "file level" or "original level", "recorded waveform" or such

The idea with "Audio gain" was to emphasize that the visuals follow the audio. The overview act like a vu-meter.

I have no string demands here. Do please consider my points and pick what works best for you.

ronso0 committed:
WaveformWidgetFactory: QStringLiteral, const ConfigKeys

ronso0 committed:
WaveformWidgetFactory: add helper getBackendFromConfig()

ronso0 commented:
> I vote for three rows, because than it is clear that the custom gain is applied to the file samples.

True, I've changed it accordingly (updated pic in description)

You're right about gain vs. factor/level, I also changed that everywhere.

daschuer reviewed (changes requested):
Looks good now. Thank you. Unfortunately the "All" gain mode is broken. (Tested this PR merged to main)

daschuer commented:
In the recording you use the "Global Gain" in the "Scrolling waveform" region. We however adjust the "Overview Waveforms". I did not expect this. 

My expectation was the the last option makes waveform follow the gain in the Mixer and the replay gain inkl. target loudness and boost value just like the scrolling one. This does not happen. 

Looking at the RGB scrolling waveforms. I am able  to shift the colours with the there "Visual Gains". The RGB overview does not follow, which creates a different visual. I think this shall not happen. How about applying the visual gain to both waveforms and in all thee gain settings?

ronso0 commented:
> In the recording you use the "Global Gain" in the "Scrolling waveform" region. We however adjust the "Overview Waveforms". I did not expect this.

Indeed slightly "unintuitive", but that's how it was since long. Run 2.5, uncheck "Normalize overview waveforms", adjust the Global ~~gain~~ scale factor (the tooltips says Gain btw).
"All" [gain] should be "Global" to clarify the situation (my mistake)

ronso0 commented:
Btw we also had and still have the issue that the "Global" factor produces different results for each waveform type.
Let's consider this one step forward and improve it in follow-ups. Maybe ReplayGain * custom factor would be the best option.

**And: I think this should go to 2.6 as this is a fixup for #14309** 
Wdyt?

daschuer commented:
> If you mean to include the deck's current pregain, I disagree.

No, pregain is unrelated. I mean the "Target loudness" and "Initial boost without ReplayGain data".  

> So the desired level is always the ReplayGain IMO.

Yes, but the replaygain target is adjustable this adjustments need to be applied. 

The waveform does not use the pregain or replaygain, it uses the total_gain 

https://github.com/mixxxdj/mixxx/blob/21a3f865615897e899423e4ec56b77bb37729817/src/waveform/renderers/waveformrenderersignalbase.cpp#L176
https://github.com/mixxxdj/mixxx/blob/21a3f865615897e899423e4ec56b77bb37729817/src/waveform/renderers/waveformwidgetrenderer.cpp#L113

Which is 
`total_gain = pregain * replaygain * ReplayGainBoost`
or 
`total_gain = pregain * DefaultBoost`
or 
`total_gain = pregain`  if replaygain is disabled 

What you need here is probably 
overview_gain = `total_gain/pregain` * all_gain (Global)` 

With this in palace we have always a waveform related to the VU-meter, but without following pregain and EQs like the scrolling waveform. This is true independent form the Replay gain preferences. 

I have initially not understand that "global" is for all waveform and the GUI with the extra gain for the file level is confusing. 

Do we really have a use case for file level if replaygain is used? This will produce tiny waveforms or overshoots dependening on the track. I would propose to just drop it. 

Than we can probably replace everything with one checkbox: 
[  ] Normalize to peak

Tooltip: 
"Uncheck to scale with the scrolling waveform but without gain and EQs."

daschuer commented:
Ah yes there is a division by zero issue whenever pregain is zereo. 
So, you may extract a reusable function from the code where total_gain is calculated and use it for the overview gain without adding pregain. Than we make sure it follow the same business logic like total_gain used for the scrolling waveforms (and audio).

ronso0 commented:
Wookay, update pushed.

I'd leave the FileLevel mode in for now.

ronso0 commented:
and squash and rebase onto 2.6 as mentioned earlier.
Wdyt?

daschuer commented:
I can confirm. It works 🎉

Would you use personally  "file level with factor"? What I cannot think of a reason. 

There is a scaling issue with the scrolling Filtered and Stacked type. They still not match to the overview. When I remove the factor 2 here: https://github.com/mixxxdj/mixxx/blob/5320334da5f3f33783871d89f59cdc862152b7a0/src/waveform/renderers/waveformwidgetrenderer.h#L148
it works. 

Regading defaults: 
Mixxx commes up with -18 LUFS replay Gain enabled. This is the setting which also allows high dynamic classic tracks. With this the waveforms look really small. 

From my experience with normal modern compressed tracks. -16 LUFS (also recommended) is fine to be always in the Green VU-meter area, but will still unlikely clip during transitions. 

In that case a Global Gain of 2.0 works perfect with all overview types. I need no headroom in the overview vissual area. Maybe this was the internal gain factor? I think we should expose it to show the truth. 


> Though, if I repeat my procedure
> 1. load a track
> 2. check the overview to find the loudest part, create a short loop
> 3. play the loop, adjust pregain to make the VU stay in the yellow region (= loud but no clipping)
> 4. in Waveform preferences, select the desired type, and adjust Global visual gain to get the same 'just not clipping' level
> 5. overview should match that

That does not work, because the pregain does only affect the scolling waveforms. You may use the Target Loudness instead. 

I can confirm it work for RGB. For filtered it is difficult, because the if the track is clipping, the sum of all bands is clipping. Not an individual peak in the waveform.

ronso0 commented:
Hmm, I have to take another look how it was in 2.5
We can of course remove the File level option if it wasn't therer previously.

daschuer commented:
The Replay Gain is now always applied. I think it shall fall back to file gain if replay gain is disabled. For users with disabled replaygain, the replaygain is not interesting. They want the file gain, to be able to estimate the manual gain the need to apply to the track.  

In 2.5 we don't have a preferences options at all for the overview waveform. It uses the file level + Global gain. This looks good for modern compressed tracks but not for high dynamic tracks. 

<img width="1396" height="341" alt="image" src="https://github.com/user-attachments/assets/e2b99e31-6bd5-4757-9ce1-1fff1adc198f" />

In this case the Global gain of 2.0 and -18 LUFS gives a good results for the moving waveform. It let's however the compressed track overview clip. This can be considered as bug. 

The following screenshot is from this branch with the ReplayGain option and the same settings like above. 

<img width="1387" height="335" alt="image" src="https://github.com/user-attachments/assets/be46f723-419e-48cb-81ee-ceb41110671d" />

We have a bit of a headroom in the Cock Sparrer track, which is the penalty for the heavy compressing during mastering. Else it looks perfect. User with only modern tracks can adjust the LUFS to -16 or more to make use of the headroom. 

Conclusion: File gain overview waeform is useless when user uses ReplayGain.

ronso0 commented:
> The Replay Gain is now always applied.

Huäh?? (german for "WTF mate?")

It worked perfectly as desired last time I built it.
Now when I rebuild, I see the wrong behavior, too. Stuck in ReplayGain mode.
Interestingly, it works again when I add some qWarning() to trace `WOverview::drawWaveformPixmap`.
All signal/slot connections are correct.

I have absolutely no clue why that happens.

ronso0 commented:
Ouuuh..

`if (m_pReplayGainEnabled->toBool()) {`
not
`if (m_pReplayGainEnabled) {` (ControlProxy)

ronso0 committed:
Overview: use namespace mixxx for OverviewType

ronso0 committed:
Overview: early return in drawWaveformPixmap(), const QRect

ronso0 committed:
Overview: scale like waveform, ie. consider ReplayGain settings

ronso0 commented:
Okay, I removed the File Level option, which leaves us with
- [ ] Normalized
- [ ] waveform "Gobal" gain + ReplayGain (if enabled)

I.e. we can handle this with the existing `[Waveform] OverviewNormalized` config option.

Thanks for your thourough review, with it I could shrink this to the essentials :+1: 

Only think left is: how to scale the overviews in the library?
(and update them when a track's ReplayGain changes)
1. always normalize to peak
2. duplicate the ReplayGain calculation in overviewRenderer (ideally move to a static helper with `ControlObject::get()`)

Currently it does normalize or use the file level.
If left a TODO for now.

ronso0 commented:
> Only think left is: how to scale the overviews in the library?

Notifying the overview chache of track gain changes is doable, it could then be passed to overview renderer, or that queries the ReplayGain from GlobalTrackCacheLocker.
But updating when ReplayGain settings changes is currently pretty elaborate (I think of DlgPrefReplayGain -> ReplayGainSettings -> MixxxMainWindow -> OverviewCache). Something like a `ReplayGainSettings` singleton might simplify this a lot.

daschuer commented:
Usingen the normalized version would be fine for me. This probably saves us from complex code.
I don't think there is much use for a gain estimation in a tiny library row.

ronso0 commented:
Are you suggesting to show **only Normalized** overviews in the library?
Would make sense because now they're changed when selecting ReplayGain but don't show the truth.

ronso0 committed:
Overview: always normalize in library

ronso0 committed:
Waveform preferences: QStringLiteral, const ConfigKeys

daschuer commented:
Code looks good, Thank you. I need to do a final manual test. 

How do we want to proceed with the different sizes in case of filtered view. The filtered view in the overview look has the same size than the RGB view (subject of this PR) . This is however not the case for the moving waveform. Here it is twice as big due to the `applyCompensation` which is now done elsewhere and it is actually an overcompensation. 

Proposal: Remove `applyCompensation` and use a 2.0 default for Global.

ronso0 commented:
This PR is only about the overviews, so please file an Issue where we can discuss the waveform gain issue.

ronso0 commented:
> Sorry, the famous close with comment button ...

In Firefox I've been using this snippet for years and it works beautifully
https://gist.github.com/ronso0/a30d1a11f4248df98cccdd3b4ccbd330
<img width="579" height="204" alt="image" src="https://github.com/user-attachments/assets/4d61f318-9a9e-4f74-b4f7-b4c884b68bf4" />

ronso0 commented:
Alright, rebased.

@daschuer please don't change the target branch.
I prefer to do this myself so that during `rebase -i` I can see which commits are actually from this branch
(I reset target to `main` and could spot them, so no big deal)

daschuer commented:
> @daschuer please don't change the target branch.

OK next time. This is now ready to go. Thank you

daschuer reviewed (approved):
Still works. LGTM
